### DISCUSSION TITLE
Current User nil with Graphql?

### DISCUSSION THREAD

**Discussion by pedrolpz99** [janko/rodauth-rails#228] [Q&A]
im trying to set the current user in my graphql controller but when i see inside byebug return nil, there is an error in my code or there is not support for graphql?

```
class ApplicationController < ActionController::API
  #include ActionController::MimeResponds

  include ActionView::Helpers::NumberHelper


  before_action :require_authentication
  def require_authentication
    rodauth.require_authentication
    Current.user = current_user
  end
  
  def current_user
    current_account = rodauth.rails_account 
    @current_user = User.find_by(account_id: current_account.id)
  end 
  helper_method :current_user

```
and in my graphql_controller current_user return nil, but in my others controllers apparently works fine, 
Do you know why that happens?
```

class GraphqlController < ApplicationController

 skip_before_action :require_authentication

  def execute
    variables = prepare_variables(params[:variables])
    query = params[:query]
    operation_name = params[:operationName]
    context = {
      current_user: current_user
    }
    result = GraphqlApiSchema.execute(query, variables: variables, context: context, operation_name: operation_name)
    render json: result
  rescue => e
    raise e unless Rails.env.development?
    handle_error_in_development e
  end
```
  
  Did I forget to set something up?
  gem "rodauth-rails", "~> 1.11"
  gem "graphql", "~> 1.11.4"
  ruby "2.7.2"

Thanks in advance!

**Comment by janko** [comment]
Are you using session middleware for your Rails API? Or are you using JWTs in `Authorization` header?

**Reply by pedrolpz99** [reply]
yes, im using JWTs Authorization header
Do you need to know more context about some specific data?
I appreciate your prompt response

**Reply by janko** [reply]
Could you share any SQL logs made by Rodauth (they will be labeled with `Sequel`) and some relevant Rodauth configuration? The only difference I see in `GraphqlController` is that it skips requiring authentication, but that shouldn't make a difference in retrieving the current account.

Note that if `current_user` returns `nil` in the code you posted, that means Rodauth was in fact able to retrieve the logged in account, because if `rodauth.rails_account` returned `nil` then `User.find_by(account_id: current_account.id)` would raise a `undefined method "id" for NilClass` error. So, my guess is that the `User` record didn't exist for this account, which is outside of Rodauth's scope.

**Reply by pedrolpz99** [reply]
Of course, sorry for the delay.
Also note that it does find the user but there seems to be a "delay" causing it to return nil in some cases

```
    9:   before_action :require_authentication
   10:   def require_authentication
   11:     rodauth.require_authentication
=> 12:     Current.user = current_user
   13:   end
   14:   
   15:   def current_user
   16:     byebug
(byebug) rodauth.rails_account
#<Account id: 91, status: "verified", email: "user@example.com", password_hash: [FILTERED]>
(byebug) current_account
#<Account id: 91, status: "verified", email: "user@example.com", password_hash: [FILTERED]>
(byebug) current_user
  CACHE User Load (0.1ms)  SELECT `users`.* FROM `users` WHERE `users`.`account_id` = 91 LIMIT 1
nil
(byebug)   ↳ app/controllers/application_controller.rb:18:in `current_user'

*** NoMethodError Exception: undefined method `id' for nil:NilClass

nil
(byebug) current_user
*** NoMethodError Exception: undefined method `id' for nil:NilClass

nil
(byebug) current_user
#<User id: 1592496726, name: "User", paternal_lastname: "Example", maternal_lastname: "", email: "user@example.com", active: true, profile_id: 1, account_id: 91>
*** NoMethodError Exception: undefined method `id' for nil:NilClass

(byebug) 
```

For rodauth app configuration just add fetures for jwt_refresh

```

class RodauthApp < Rodauth::Rails::App
  # primary configuration
  configure RodauthMain

  # secondary configuration
  plugin :rodauth do
    enable  :jwt_refresh,:internal_request, :create_account
    jwt_access_token_period 500
    jwt_refresh_token_deadline_interval {30.minutes}
    allow_refresh_with_expired_jwt_access_token? true
    expired_jwt_access_token_status 401

    before_refresh_token do
      current_account = account.to_hash
      user = User.joins(:account, :franchise).find_by(account_id: current_account[:id])
    end
```

To configure rodauth main, sign the token with some data from my users, I only removed verify_account feture

```
class RodauthMain < Rodauth::Rails::Auth
  configure do

  jwt_session_hash do
      original_token = super()
      if original_token[:account_id] then
        current_account = User.includes(:account, ...)
                     .find_by(account_id: original_token[:account_id])
      
        user_json = {
          id:current_account.id,
          email:current_account.email,
          name: current_account.name,
           ...
           }

        original_token.merge(:user=>user_json)
      else
        original_token
      end
    end

```

**Reply by janko** [reply]
> Also note that it does find the user but there seems to be a "delay" causing it to return nil in some cases

I'm not sure what kind of a delay this is, could be some caching issue? Are you using the latest version of `sequel-activerecord_connection` gem? It takes care of clearing Active Record's query cache after Rodauth's SQL statements.

If that's already the case, then I would need you to create a sample Rails application that demonstrates the problem, so that I can run it locally and debug it.
